Dalamud/Game/AddonLifecycle/AddonLifecycle.cs

```diff
@@ -0,0 +1,168 @@
+﻿using System;
+using System.Runtime.InteropServices;
+
+using Dalamud.Hooking;
+using Dalamud.IoC;
+using Dalamud.IoC.Internal;
+using Dalamud.Logging;
+using Dalamud.Plugin.Services;
+using FFXIVClientStructs.FFXIV.Component.GUI;
+
+namespace Dalamud.Game.AddonLifecycle;
+
+/// <summary>
+/// This class provides events for in-game addon lifecycles.
+/// </summary>
+[InterfaceVersion("1.0")]
+[ServiceManager.EarlyLoadedService]
+internal unsafe class AddonLifecycle : IDisposable, IServiceType, IAddonLifecycle
+{
+    private readonly AddonLifecycleAddressResolver address;
+    private readonly Hook<AddonSetupDelegate> onAddonSetupHook;
+    private readonly Hook<AddonFinalizeDelegate> onAddonFinalizeHook;
+    
+    [ServiceManager.ServiceConstructor]
+    private AddonLifecycle(SigScanner sigScanner)
+    {
+        this.address = new AddonLifecycleAddressResolver();
+        this.address.Setup(sigScanner);
+
+        this.onAddonSetupHook = Hook<AddonSetupDelegate>.FromAddress(this.address.AddonSetup, this.OnAddonSetup);
+        this.onAddonFinalizeHook = Hook<AddonFinalizeDelegate>.FromAddress(this.address.AddonSetup, this.OnAddonFinalize);
+    }
+
+    [UnmanagedFunctionPointer(CallingConvention.ThisCall)]
+    private delegate nint AddonSetupDelegate(AtkUnitBase* addon);
+
+    [UnmanagedFunctionPointer(CallingConvention.ThisCall)]
+    private delegate void AddonFinalizeDelegate(AtkUnitManager* unitManager, AtkUnitBase** atkUnitBase);
+    
+    /// <inheritdoc/>
+    public event Action<IAddonLifecycle.AddonArgs>? AddonPreSetup;
+    
+    /// <inheritdoc/>
+    public event Action<IAddonLifecycle.AddonArgs>? AddonPostSetup;
+    
+    /// <inheritdoc/>
+    public event Action<IAddonLifecycle.AddonArgs>? AddonPreFinalize;
+    
+    /// <inheritdoc/>
+    public event Action<IAddonLifecycle.AddonArgs>? AddonPostFinalize;
+    
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        this.onAddonSetupHook.Dispose();
+        this.onAddonFinalizeHook.Dispose();
+    }
+    
+    [ServiceManager.CallWhenServicesReady]
+    private void ContinueConstruction()
+    {
+        this.onAddonSetupHook.Enable();
+        this.onAddonFinalizeHook.Enable();
+    }
+
+    private nint OnAddonSetup(AtkUnitBase* addon)
+    {
+        try
+        {
+            this.AddonPreSetup?.Invoke(new IAddonLifecycle.AddonArgs { Addon = (nint)addon });
+        }
+        catch (Exception e)
+        {
+            PluginLog.Error(e, "[AddonLifecycle] Exception in OnAddonSetup pre-setup invoke.");
```

Comment:
Please use `ModuleLog` here. e.g. at the top of the class:

```cs
private static readonly ModuleLog Log = new("AddonLifecycle");
```

and then call `Log.Thing()`.

Applies to all PluginLog uses.